corehq/apps/users/tests/test_tasks.py

```diff
@@ -176,3 +184,76 @@ def _send_case_to_es(
 
         case_search_adapter.index(case, refresh=True)
         return case
+
+
+@mock.patch.object(UserReportingMetadataStaging, 'process_record')
+class TestProcessReportingMetadataStaging(TestCase):
+
+    def test_record_is_deleted_if_processed_successfully(self, mock_process_record):
+        record = UserReportingMetadataStaging.objects.create(user_id=self.user._id, domain='test-domain')
+        self.assertTrue(UserReportingMetadataStaging.objects.get(id=record.id))
+
+        _process_reporting_metadata_staging()
+
+        self.assertEqual(mock_process_record.call_count, 1)
+        self.assertEqual(UserReportingMetadataStaging.objects.all().count(), 0)
+
+    def test_record_is_not_deleted_if_not_processed_successfully(self, mock_process_record):
+        record = UserReportingMetadataStaging.objects.create(user_id=self.user._id, domain='test-domain')
+        mock_process_record.side_effect = Exception
+
+        with self.assertRaises(Exception):
+            _process_reporting_metadata_staging()
+
+        self.assertEqual(mock_process_record.call_count, 1)
+        self.assertTrue(UserReportingMetadataStaging.objects.get(id=record.id))
+
+    def test_process_record_is_retried_if_resource_conflict_raised(self, mock_process_record):
+        # Simulate the scenario where the first attempt to process a record raises ResourceConflict
+        # but the next attempt succeeds
+        mock_process_record.side_effect = [ResourceConflict, None]
+        UserReportingMetadataStaging.objects.create(user_id=self.user._id, domain='test-domain')
+
+        _process_reporting_metadata_staging()
+
+        self.assertEqual(mock_process_record.call_count, 2)
+        self.assertEqual(UserReportingMetadataStaging.objects.all().count(), 0)
+
+    def test_subsequent_records_are_not_processed_if_exception_raised(self, mock_process_record):
+        mock_process_record.side_effect = [Exception, None]
+        UserReportingMetadataStaging.objects.create(user_id=self.user._id, domain='test-domain')
+        UserReportingMetadataStaging.objects.create(user_id=self.user._id, domain='test-domain')
+
+        with self.assertRaises(Exception):
+            _process_reporting_metadata_staging()
+
+        self.assertEqual(mock_process_record.call_count, 1)
+        self.assertEqual(UserReportingMetadataStaging.objects.all().count(), 2)
+
+    def setUp(self):
+        super().setUp()
+        self.user = CommCareUser.create('test-domain', 'test-username', 'qwer1234', None, None)
+        self.addCleanup(self.user.delete, 'test-domain', deleted_by=None)
+
+
+@mock.patch.object(UserReportingMetadataStaging, 'process_record')
+class TestProcessReportingMetadataStagingTransaction(TransactionTestCase):
+    """
+    TransactionTestCase is much slower than TestCase, but needed to use new_db_connection
```

Comment:
Why is it a requirement to use `TransactionTestCase` with `new_db_connection`? [This test](https://github.com/dimagi/commcare-hq/blob/81f3a488487ec6e2aa60c33c8c0d795c0d1ac9e8/corehq/apps/fixtures/tests/test_upload.py#L656) doesn't do that.

Reply:
It is for `select_for_update`, not `new_db_connection`. From [django docs](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#select-for-update)
> Warning
>
> Although select_for_update() normally fails in autocommit mode, since [TestCase](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#django.test.TestCase) automatically wraps each test in a transaction, calling select_for_update() in a TestCase even outside an [atomic()](https://docs.djangoproject.com/en/3.2/topics/db/transactions/#django.db.transaction.atomic) block will (perhaps unexpectedly) pass without raising a TransactionManagementError. To properly test select_for_update() you should use [TransactionTestCase](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#django.test.TransactionTestCase).

Reply:
I see. The docstring says "needed to use new_db_connection"

However, after reading those docs I still have my doubts that `TransactionTestCase` is required. I think you could do the `select_for_update` in the `new_db_connection` block, and all should be fine since that connection will be closed (and its transaction ended, which terminates `select_for_update`) when the block exits.

Reply:
Oh my bad! Thanks for catching that. Yeah I agree with you although I think the `select_for_update` needs to be outside of the `new_db_connection`. The purpose of the `new_db_connection` is to ensure the method under test uses a different db connection from the one that holds the lock on a record when attempting to fetch records from the db. But as far as TestCase vs TransactionTestCase, since a TestCase wraps the test in a transaction, I would assume running `select_for_update` in the test and opening a new db connection would work fine, but I've found that in this scenario, the method under tests is unable to fetch _any_ from the database which I don't quite understand. In a TransactionTestCase this doesn't happen. Any ideas on what might be the problem?

Reply:
Yes, the transaction management in `TestCase` is a little more complicated than that. It starts a transaction during `setUpClass`, then makes a SAVEPOINT before each test, and it rolls back to and releases the SAVEPOINT after each test. Finally, it rolls back the entire transaction in `tearDownClass` after all tests have completed.

`ROLLBACK TO SAVEPOINT` does not release locks acquired by `SELECT FOR UPDATE` because [that would violate the ACID properties of the transaction](https://www.postgresql.org/message-id/1154639749.24639.22.camel%40dogma.v10.wvs). Therefore, any effects of `SELECT FOR UPDATE` in one test will be visible to other tests in the same class that are run later.

> I think the `select_for_update` needs to be outside of the `new_db_connection`. The purpose of the `new_db_connection` is to ensure the method under test uses a different db connection from the one that holds the lock on a record

Ah yes, I get it. We could work around this with another context manager that restores the original connection once the `select_for_update` has run, but before closing the "new connection". This would be pretty easy to write, and it would work something like this:

```py
with new_db_connection() as cn:
    _ = UserReportingMetadataStaging.objects.select_for_update().get(pk=record.id)
    with cn.restore_previous_connection():
        _process_reporting_metadata_staging()
```

Let me know if you want help writing that.

Another way we could solve this problem without implementing `restore_previous_connection` would be to put this test in its own `TestCase` and document that it should be the only test in that class. That would be a bit hacky, but easy enough.

Reply:
@millerdev and I paired on this and came up with this solution: https://github.com/dimagi/commcare-hq/pull/33745/commits/00ab97f2d0dd6bb7cef334fb3f582693891a481c. 

It isn't the prettiest, but allows us to continue to use a TestCase which is faster and has a safer cleanup mechanism than TransactionTestCase.